engine/src/main/java/org/destinationsol/game/SolObject.java

```diff
@@ -16,41 +16,187 @@
+     * when such is applicable. Note that this method is also called when an object is due to be removed for
+     * optimization purposes, and thus you need to explicitly check for {@link #shouldBeRemoved(SolGame)} if you have
+     * some code specific to destruction of object.
+     *
+     * @param game Game this object belongs to.
+     */
     void onRemove(SolGame game);
 
+    /**
+     * Called whenever this object should receive damage.
+     * <p>
+     * Whenever the object is hit by some projectile or other sort of damage dealing thing, this method is called on the
+     * object to allow it to perform its damage handling. This usually means just subtracting the damage from object's
+     * health and playing a hit sound, but if the object has no health pool or should be otherwise indestructible, or
+     * invincible against some types of damage, this method can be freely left blank.
+     *
+     * @param dmg      Damage the object receives.
+     * @param game     Game this object belongs to.
+     * @param position Position the object was hit at, if hit by point-based damage. Null if not applicable, such as fire.
+     * @param dmgType  Type of the damage object receives.
+     */
     void receiveDmg(float dmg, SolGame game, @Nullable Vector2 position, DmgType dmgType);
 
+    /**
+     * Denotes whether this object is affected by gravity, this is usually constant.
+     *
+     * @return True if object is affected by gravity, false otherwise.
+     */
     boolean receivesGravity();
 
+    /**
+     * Applies external force to object.
+     * <p>
+     * This method usually just scales the {@code force} by objects mass if {@code acc} is set, and then passes the
+     * force to object's internal {@link Body}.
+     *
+     * @param force Force to apply to the object.
+     * @param game  Game this object belongs to.
+     * @param acc   Whether this objects is accelerating by the force, such as in gravity.
+     */
     void receiveForce(Vector2 force, SolGame game, boolean acc);
 
+    /**
+     * Returns object's position.
+     * <p>
+     * This method usually just returns the vector retrieved from object's internal {@link Body}.
+     *
+     * @return Position of the object.
+     */
     Vector2 getPosition();
 
+    /**
+     * Called when an object is too far from player/camera, and is thus to be converted to {@link FarObject}.
+     * <p>
+     * This method is used for optimization of game's objects, by converting to their more resource friendly versions
+     * when they get too far from player/camera. The {@code FarObject} returned by this method should be usable to get
+     * copy of this object back in reverse. If null is returned by this method, this object is designated to be removed
+     * when it gets too far from player/camera, and instead of recreating from {@code FarObject}, to be generated again.
```

Comment:
`* when it gets too far from the player/camera and has to be regenerated manually.`

I may be wrong here, so please correct me if the object does not actually have to be regenerated manually in this case.

Reply:
What does regenerated mmanually mean?

Reply:
By "regenerated manually", I mean that the engine does not "optimize" this `SolObject` by converting it to a `FarObject`, it is removed and has to be created again. That is what I could gather from your description. Is that what really happens? If not, we might have to rephrase this to accurately reflect what will happen if `toFarObject` returns null.

Reply:
Then why the word "manually"? With your suggested wording, it would make it seem to me that you have to add some special code somewhere to allow the object to be generated new in an area that has been already generated before.

Reply:
I am not sure I fully understand your last comment, but yes, I do want to make it seem that we have to add code to allow the object to be regenerated new in an area where it has been already generated before. As far as I can understand, that is what `and instead of recreating from {@code FarObject}, to be generated again` means. If this is not your intention/not actually what happens, you might want to consider rewording that line to reflect what actually happens. If you'd like, you can post a description of what actually happens and we can come up with something together.

Reply:
Objects that have their respective FarObjects are usually generated when the system is generated, and are just loaded/unloaded as needed. For other objects, there is a generateChunk() that is being called when player gets close to a chunk that has not been loaded, this function creates all SolObjects in the chunk without persistance, when the player gets too far again, these objects are destroyed, and chunk unloaded. Upon next coming close towards the chunk,, the generateChunk() is executed again.

Reply:
Adrijaned, thank you for that detailed, well written description! It certainly helps clear all confusion. Keeping that in mind, I accept your suggestion that we drop the word "manually", and change the clause to `when it gets too far from the player/camera and is regenerated from scratch (without persistence)`, which hopefully includes all the required details. What do you think?